connect_timeout + read_timeout

kp-cat opened the pull request:
### Describe the purpose of your pull request

Introduce connect_timeout + read_timeout on the interface.
Logging the timeout values when a timeout exception occurs.

### Requirement checklist (only if applicable)

- [x] I have covered the applied changes with automated tests.
- [ ] I have executed the full automated test set against my changes.
- [ ] I have validated my changes against all supported platform versions.
- [x] I have read and accepted the [contribution agreement](https://github.com/configcat/legal/blob/main/contribution-agreement.md).

kp-cat committed:
connect_timeout + read_timeout

randycoulman commented on lib/config_cat/config_fetcher.ex:
```diff
@@ -101,6 +103,14 @@ defmodule ConfigCat.CacheControlConfigFetcher do
+            Logger.error(
+              "Request timed out. Timeout values: [connect: #{state.connect_timeout}ms, read: #{state.read_timeout}ms]"
+            )
+        end
```
Would it be better to move this additional logging down to `log_error` so that all error logging is contained in a single function?

#optional

randycoulman reviewed (approved):
LGTM! One optional suggestion below.

randycoulman commented:
I think this is failing CI because of changes to the Elixir formatter between versions. We probably need to take the formatting check out of CI, or only run it on the most recent version of Elixir or something like that. It also looks like we should update our CI matrix to include newer versions of Elixir.

randycoulman commented:
@kp-cat See #73 for what should fix the CI issues. If you want to approve that, I can merge it and then you can rebase on it to get this PR green.

kp-cat committed:
Merge remote-tracking branch 'origin/main' into timeout

kp-cat commented:
Hi @randycoulman, I merged your PR, thank you.

kp-cat committed:
move timeout error log into log_error function

randycoulman commented on lib/config_cat.ex:
```diff
@@ -172,6 +186,8 @@ defmodule ConfigCat do
           | {:data_governance, data_governance()}
           | {:http_proxy, String.t()}
+          | {:connect_timeout, Integer.t()}
+          | {:read_timeout, Integer.t()}
```
I missed this in my earlier review. These types (and below) should be `integer()` or `non_neg_integer()` probably. There is no such thing as `Integer.t()` in Elixir.

This is the root cause of the current CI failures.

See https://hexdocs.pm/elixir/1.13/typespecs.html for more details.

randycoulman commented on lib/config_cat/config_fetcher.ex:
```diff
@@ -216,9 +221,15 @@ defmodule ConfigCat.CacheControlConfigFetcher do
     Logger.error("Double-check your SDK Key at https://app.configcat.com/sdkkey.")
     Logger.error("Failed to fetch configuration from ConfigCat: #{inspect(error)}")
+
+    if error == {:error, %HTTPoison.Error{reason: :timeout}} do
```
This conditional probably won't work like you expect because it's doing a strict equality comparison, and an `HTTPoison.Error` struct almost certainly has more fields than just the one being checked here.

The earlier version with the `case` statement would have used pattern matching instead of direct equality comparison and that would have worked.

Options:
- You could go back to the `case` version you had earlier
- You could look at using [`Kernel.match?`](https://hexdocs.pm/elixir/1.13/Kernel.html#match?/2) to do the comparison
- You could extract a private helper function, something like:

```elixir
defp log_timeout({:error, %HTTPoison.Error{reason: :timeout}}) do
  Logger.error(...)
end

defp log_timeout(_error), do: :ok
```

randycoulman reviewed (commented):
Looks like some changes needed to fix this up a bit and to get CI passing again.

kp-cat committed:
use non_neg_integer type for timeouts

kp-cat committed:
timeout error handling fix

Commit:
💚 Fix failing tests

Many tests were over-specifying the options being passed `API.get`, so the tests were failing because of the new timeout options always being passed.

Also, add more tests around the timeout settings in ConfigFetcher.

Two of the failures were because the new timeout logging code wasn't handling all of the possible cases.

kp-cat commented:
Thanks @randycoulman for the fixes 💯
